manager/orchestrator/replicatedjob/orchestrator.go

```diff
@@ -45,6 +52,166 @@ func (o *Orchestrator) Run(ctx context.Context) {
+	// all integers from 0 to (MaxConcurrent - 1), with a boolean indicating if
+	// that slot is occupied. This slot handling code is much simpler than the
+	// code for replicated services, because we don't need to worry about
+	// restarts.
+	slots := map[uint64]bool{}
+	for _, task := range tasks {
+		// we only care about tasks from this job iteration. tasks from the
+		// previous job iteration are not important
+		// TODO(dperny): we need to stop any running tasks from older job
+		// iterations.
+		if task.JobIteration != nil && task.JobIteration.Index == jobVersion {
+			if task.Status.State == api.TaskStateCompleted {
+				completeTasks++
+			}
+
+			// any tasks that are created and running but not yet terminal are
+			// considered "running tasks" for our purpose, because they don't
+			// need to be created.
+			if task.Status.State <= api.TaskStateRunning && task.DesiredState == api.TaskStateCompleted {
+				runningTasks++
+				slots[task.Slot] = true
+			}
+		}
+	}
+
+	// now that we have our counts, we need to see how many new tasks to
+	// create. this number can never exceed MaxConcurrent, but also should not
+	// result in us exceeding TotalCompletions. first, get these numbers out of
+	// the service spec.
+	rj := service.Spec.GetReplicatedJob()
+
+	// possibleNewTasks gives us the upper bound for how many tasks we'll
+	// create. also, ugh, subtracting uints. there's no way this can ever go
+	// wrong.
+	possibleNewTasks := rj.MaxConcurrent - runningTasks
+
+	// allowedNewTasks is how many tasks we could create, if there were no
+	// restriction on maximum concurrency. This is the total number of tasks
+	// we want completed, minus the tasks that are already completed, minus
+	// the tasks that are in progress.
+	//
+	// seriously, ugh, subtracting unsigned ints. totally a fine and not at all
+	// risky operation, with no possibility for catastrophe
+	allowedNewTasks := rj.TotalCompletions - completeTasks - runningTasks
+
+	// the lower number of allowedNewTasks and possibleNewTasks is how many we
+	// can create. we'll just use an if statement instead of some fancy floor
+	// function.
+	actualNewTasks := allowedNewTasks
+	if possibleNewTasks < allowedNewTasks {
+		actualNewTasks = possibleNewTasks
+	}
+
+	// this check might seem odd, but it protects us from an underflow of the
+	// above subtractions, which, again, is a totally impossible thing that can
+	// never happen, ever, obviously.
+	if actualNewTasks > rj.TotalCompletions {
+		return fmt.Errorf(
+			"uint64 underflow, we're not going to create %v tasks",
+			actualNewTasks,
+		)
+	}
+
+	// finally, we can create these tasks. do this in a batch operation, to
+	// avoid exceeding transaction size limits
+	err := o.store.Batch(func(batch *store.Batch) error {
+		for i := uint64(0); i < actualNewTasks; i++ {
+			if err := batch.Update(func(tx store.Tx) error {
+				// find an unoccupied slot number that we can use, same as with
```

Comment:
Would it be worth it to build once and for all a slice of all unoccupied slots, instead of iterating over the map for every new job here?

Reply:
I don't think so, because then we'd have to carefully keep track of which slots opened up and when, when this operation doesn't have all that much overhead, and the overhead pales in comparison to the overhead of doing a store update.

Reply:
Fair enough, thanks :)